File: Solutions/1007_Minimum_Domino_Rotations_For_Equal_Row/1007_Minimum_Domino_Rotations_For_Equal_Row.go
//Most common
func minDominoRotations(A []int, B []int) int {
    rotations := 1 << 63-1
    
    rotations = min(rotations, getDominoRot(A, B, A[0]))
    rotations = min(rotations, getDominoRot(B, A, B[0]))
    rotations = min(rotations, getDominoRot(A, B, B[0]))
    rotations = min(rotations, getDominoRot(B, A, A[0]))
    
    if rotations == 1 <<63-1 {
        return -1
    }
    return rotations
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func getDominoRot(a, b []int, val int) int {
    if len(a) == 0 || len(b) == 0 {
        return 0
    }
    
    countRot := 0
    for i := range a {
        if a[i] == val {
            continue
        } else if b[i] == val {
            countRot++
        } else {
            return 1 <<63 -1
        }
    }
    
    
    return countRot
}

//Third fastest
func minDominoRotations(A []int, B []int) int {
    res := helper(A, B, A[0])
    if res != -1 || A[0] == B[0]{
        return res
    }
    return helper(A, B, B[0])
}

func helper(A []int, B []int, x int) int {
    i := 0
    rotCountA := 0
    rotCountB := 0
    for i < len(A){
        if A[i] != x && B[i] != x{
            return -1
        }
        if A[i] != x {
            rotCountA++
        }else if B[i] != x {
            rotCountB++
        }
        i++
    } 
    if rotCountA < rotCountB {
        return rotCountA
    }
    return rotCountB
}

//Second fastest
func minDominoRotations(A []int, B []int) int {
    a := A[0]
    b := B[0]
    
    countATop := 1
    countBTop := 0
    
    countABottom := 0
    countBBottom := 1
    for i := 1; i < len(A); i++ {
        if a != A[i] && a != B[i] {
            a = -1
        }
        
        if b != A[i] && b != B[i] {
            b = -1
        }
        
        if a == A[i] {
            countATop++
        }
        
        if b == A[i] {
            countBTop++
        }
        
        if a == B[i] {
            countABottom++
        }
        
        if b == B[i] {
            countBBottom++
        }
        
        if a == -1 && b == -1 {
            return -1
        }
    }
    
    if a == b {
        return len(A) - max(max(countATop, countABottom), max(countBTop, countBBottom))
    }
    
    if a > b {
        return len(A) - max(countATop, countABottom)
    }
    return len(A) - max(countBTop, countBBottom)
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

//Fastest
func minDominoRotations(A []int, B []int) int {

	var aCtr = make([]int, 8)
	var bCtr = make([]int, 8)
	var sameCtr = make([]int, 8)

	for i := range A {
		if A[i] == B[i] {
			sameCtr[A[i]]++
		}

		aCtr[A[i]]++
		bCtr[B[i]]++
	}

	min := 99999

	for i := 1; i <= 6; i++ {

		if aCtr[i]+bCtr[i]-sameCtr[i] == len(A) {

			min = mn(min, mn(aCtr[i]-sameCtr[i], bCtr[i]-sameCtr[i]))

		}
	}

	if min == 99999 {
		return -1
	}

	return min

}

func mn(a, b int) int {

	if a < b {
		return a
	}

	return b

}

File: Solutions/1007_Minimum_Domino_Rotations_For_Equal_Row/1007_Minimum_Domino_Rotations_For_Equal_Row.py
#Third fastest
class Solution:
    def minDominoRotations(self, A: List[int], B: List[int]) -> int:
        possible_sol = {A[0],B[0]}
        
        for i in possible_sol:
            A_swap = 0
            B_swap = 0
            flag = 0
            for j in range(len(A)):
                if A[j] == B[j] == i:
                    pass
                elif A[j] == i:
                    B_swap += 1
                elif B[j] == i:
                    A_swap += 1
                else:
                    flag = 1
                    break
                    
            if flag == 0:
                return min(A_swap,B_swap)
        return -1

#Second fastest
class Solution:        
    def minDominoRotations(self, A: List[int], B: List[int]) -> int:
        def check(x):
            """
            Return min number of swaps 
            if one could make all elements in A or B equal to x.
            Else return -1.
            """
            # how many rotations should be done
            # to have all elements in A equal to x
            # and to have all elements in B equal to x
            rotations_a = rotations_b = 0
            for i in range(n):
                # rotations coudn't be done
                if A[i] != x and B[i] != x:
                    return -1
                # A[i] != x and B[i] == x
                elif A[i] != x:
                    rotations_a += 1
                # A[i] == x and B[i] != x    
                elif B[i] != x:
                    rotations_b += 1
            # min number of rotations to have all
            # elements equal to x in A or B
            return min(rotations_a, rotations_b)
    
        n = len(A)
        rotations = check(A[0]) 
        # If one could make all elements in A or B equal to A[0]
        if rotations != -1 or A[0] == B[0]:
            return rotations 
        # If one could make all elements in A or B equal to B[0]
        else:
            return check(B[0])

#Fastest
class Solution:        
    def minDominoRotations(self, A: List[int], B: List[int]) -> int:
        assert(len(A)>=2)
        def check(x):
            """
            Return min number of swaps 
            if one could make all elements in A or B equal to x.
            Else return -1.
            """
            # how many rotations should be done
            # to have all elements in A equal to x
            # and to have all elements in B equal to x
            rotations_a = rotations_b = 0
            for i in range(n):
                # rotations coudn't be done
                if A[i] != x and B[i] != x:
                    return -1
                # A[i] != x and B[i] == x
                elif A[i] != x:
                    rotations_a += 1
                # A[i] == x and B[i] != x    
                elif B[i] != x:
                    rotations_b += 1
            # min number of rotations to have all
            # elements equal to x in A or B
            return min(rotations_a, rotations_b)
    
        n = len(A)
        rotations = check(A[0]) 
        # If one could make all elements in A or B equal to A[0]
        if rotations != -1:
            return rotations 
        # If one could make all elements in A or B equal to B[0]
        else:
            return check(B[0])